manim/utils/qhull.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python
+from __future__ import annotations
+
+import numpy as np
+
+
+class Point:
+    def __init__(self, coordinates):
+        self.coordinates = coordinates
+
+    def __hash__(self):
+        return hash(self.coordinates.tobytes())
+
+    def __eq__(self, other):
+        return np.array_equal(self.coordinates, other.coordinates)
+
+
+class SubFacet:
+    def __init__(self, coordinates):
+        self.coordinates = coordinates
+        self.points = frozenset(Point(c) for c in coordinates)
+
+    def __hash__(self):
+        return hash(self.points)
+
+    def __eq__(self, other):
+        return self.points == other.points
+
+
+class Facet:
+    def __init__(self, coordinates, normal=None, internal=None):
+        self.coordinates = coordinates
+        self.center = np.mean(coordinates, axis=0)
+        self.normal = self.compute_normal(internal)
+        self.subfacets = frozenset(
+            SubFacet(np.delete(self.coordinates, i, axis=0))
+            for i in range(self.coordinates.shape[0])
+        )
+
+    def compute_normal(self, internal):
+        centered = self.coordinates - self.center
+        _, _, vh = np.linalg.svd(centered)
+        normal = vh[-1, :]
+        normal /= np.linalg.norm(normal)
+
+        if np.dot(normal, self.center - internal) < 0:
+            normal *= -1
+
+        return normal
+
+    def __hash__(self):
+        return hash(self.subfacets)
+
+    def __eq__(self, other):
+        return self.subfacets == other.subfacets
+
+
+class Horizon:
+    def __init__(self):
+        self.facets = set()
+        self.boundary = []
+
+
+class QuickHull:
+    def __init__(self, tolerance=1e-5):
+        self.facets = []
+        self.removed = set()
+        self.outside = {}
+        self.neighbors = {}
+        self.unclaimed = None
+        self.internal = None
+        self.tolerance = tolerance
```

Comment:
It would be nice if you added docstrings and typings for parameters, returns and attributes (where necessary), both in this file and `polylabel.py`.

It allows for people to hover on an attribute or method (in Visual Studio Code, for example) and get hints on what the type of the attribute is supposed to be, what the parameters of a method are expected to be, and what you should expect to get as a result when calling a method.

It makes dev debugging much easier, and it allows our type checkers to detect errors.

I suggest that you take a look at: https://docs.manim.community/en/stable/contributing/docs.html

For example, it could be something like this (I guessed the data types, so maybe I'm wrong on some of these):

```suggestion
class QuickHull:
    """Description of a QuickHull.
    
    Parameters
    ----------
    tolerance
        Description of what the tolerance is.
        
    Attributes
    ----------
    facets
        List of facets.
    removed
        A set of removed facets.
    etc...
    """
    def __init__(self, tolerance: float = 1e-5) -> None:
        self.facets: list[Facet] = []
        self.removed: set[Facet] = set()
        self.outside: dict[Facet, tuple[Point3D_Array | None, Point3D | None]] = {}
        self.neighbors: dict[SubFacet, Facet] = {}
        self.unclaimed: Point3D_Array | None = None
        self.internal: Point3D | None = None
        self.tolerance = tolerance
```

Reply:
Yes, I can add docstrings and types to this! Actually, the [original code](https://github.com/JayGupta797/convex-hull) was already strongly typed, so it should be a matter of adding those back with some minor adjustments. Unfortunately, I can't use `Point3D` because the algorithm extends to arbitrary dimensions (thought in practice it is very slow beyond 7).

Reply:
I see. Manim only works with points of dimension up to 3, so `Point3D` could be fine for this anyways... but we can always add `PointND` and `PointND_Array` to `manim.typing` if necessary (we currently do have `VectorND` as a type alias, but that's for directions rather than positions).

Reply:
I see! The only other issue I can think of at the moment is handling the 2 and 3 dimensional cases with `Point3D` since the former would take the form of `[x y 0]` leading to undesirable behavior. I think the simplest solution is to use `np.ndarray` for typing here.